File: 0680-valid-palindrome-ii/0680-valid-palindrome-ii.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        left=0
        right=len(s)-1
        
        def is_pal(left,right):
            l = left
            r = right
            while l<r:
                if s[l]!=s[r]:
                    return False
                l+=1
                r-=1
            return True

        while left<right:
            if s[left]!=s[right]:
                return is_pal(left+1,right) or is_pal(left,right-1)
            left+=1
            right-=1
        
        return True

File: 0680-valid-palindrome-ii/0680-valid-palindrome-ii.go
func validPalindrome(s string) bool {
	left := 0
	right := len(s) - 1

	for left < right {
		// If characters match, just keep moving inward
		if s[left] == s[right] {
			left++
			right--
		} else {
			// Mismatch found! We use our "one deletion" chance here.
			// We check if the substring is valid by skipping 'left' OR skipping 'right'.
			return isPalindromeRange(s, left+1, right) || isPalindromeRange(s, left, right-1)
		}
	}
	return true
}

// Helper function to check if a specific range of the string is a palindrome
func isPalindromeRange(s string, left, right int) bool {
	for left < right {
		if s[left] != s[right] {
			return false
		}
		left++
		right--
	}
	return true
}